exchange/exchange.go

```diff
@@ -305,6 +312,131 @@ func (e *exchange) buildBidResponse(ctx context.Context, liveAdapters []openrtb_
 	return bidResponse, err
 }
 
+func applyCategoryMapping(requestExt openrtb_ext.ExtRequest, seatBids *map[openrtb_ext.BidderName]*pbsOrtbSeatBid, categoriesFetcher stored_requests.CategoryFetcher) (map[string]string, error) {
+	res := make(map[string]string)
+
+	//If includebrandcategory is present in ext then CE feature is on.
+	if requestExt.Prebid.Targeting == nil {
+		return res, nil
+	}
+	brandCatExt := requestExt.Prebid.Targeting.IncludeBrandCategory
+
+	//If includebrandcategory is present in ext then CE feature is on.
+	if brandCatExt == (openrtb_ext.ExtIncludeBrandCategory{}) {
+		return res, nil //if not present continue the existing processing without CE.
+	}
+
+	//if includebrandcategory present but primaryadserver/publisher not present then error out the request right away.
+	primaryAdServer, err := getPrimaryAdSetver(brandCatExt.PrimaryAdServer) //1-Freewheel 2-DFP
+	if err != nil {
+		return res, err
+	}
+
+	publisher := brandCatExt.Publisher
+
+	seatBidsToRemove := make([]openrtb_ext.BidderName, 0)
+
+	for bidderName, seatBid := range *seatBids {
+		bidsToRemove := make([]int, 0)
+		for bidInd := range seatBid.bids {
+			bid := seatBid.bids[bidInd].bid
+			var tempRespBidExt ResponseBidExt
+			if len(bid.Ext) > 0 {
+				fmt.Println(string(bid.Ext))
+				if err := json.Unmarshal(bid.Ext, &tempRespBidExt); err != nil {
+					return res, err
+				}
+				//if the bid already consists bid.ext.prebid.category then it implies that adapter has returned the final Ad sever category.
+				//if not then look at the bid.cat[0] and translate it to the mapped Ad server category
+				if len(tempRespBidExt.Prebid.Targeting.HbPbCatDur) != 0 {
```

Comment:
Were we expecting the adapters to provide the combined pb_cat_dur string when they are returning the primary ad server category directly, or just the category? I thought the latter and the combining of these fields would be done setting the targeting data, but that was probably just my assumption I don't know if we ever specifically decided one way or the other.

Reply:
@jmaynardxandr Jason, I implemented this check to meet specification requirements, described in [Project Sahara wiki page](https://corpwiki.appnexus.com/pages/viewpage.action?spaceKey=CT&title=Project+Sahara+Prebid+Server+Tech+Spec). 
We can ask Hans about this explicitly.

Reply:
Yeah sounds good, we'll discuss with Hans tomorrow. My understanding was some field holds the translated category, if that field's not set by the adapter then we look up the IAB category and set it from there. Then the setTargeting() function combines the price bucket, translated category, and duration into the final key/value pair. We also should make sure to ask Hans about the lack of duration in the bid request tomorrow. After we figure this out everything else looks good to me, great job!

Reply:
Still need a little work here. First thing is "tempRespBidExt" is not being unmarshalled to after this update so it won't catch the case where the adapter returns a primary ad server category. Second rename the "HbPbCatDur" to something indicating it holds just the primary ad server category ("PrimaryCategory" maybe?). Third is we agreed the Duration should be a field in the prebid extension so should have a tempRespBidExt.Prebid.Duration field we can use here, and the appnexus.go adapter should copy the video duration into that field so you don't have to unmarshal the bidder extension here.

Reply:
@jmaynardxandr Jason, thank you for the comment!
1)Yes, sure, I don't know why did I delete Unmarshalling logic :)
2)Renamed
3)Did we agreed about this? If so, can we have different formats of bid extension? We build targeting data after categories logic. I added code to handle different type of bid extension. We may have this format (with prebid and duration in it): 
`{"prebid": {"duration": 50, "targeting": {"hb_bidder": "appnexus","hb_pb": "20.00","hb_pb_cat_dur": "Sports","hb_size": "1x1"},"type": "video"},"bidder": {"appnexus":{"brand_id":1,"auction_id":4797544351646081865,"bidder_id":2,"bid_ad_type":1,"creative_info":{"video":{"duration":30,"mimes":["video\/mp4"]}}} }}`
And this format (current format): 
`{"appnexus"{"brand_id":1,"auction_id":4797544351646081865,"bidder_id":2,"bid_ad_type":1,"creative_info":{"video":{"duration":30,"mimes":["video\/mp4"]}}}}`
Also I added test scenario for this.
Please advice.

Reply:
I squashed commits to keep commits history clear

Reply:
Ok after getting more into the code involved having the adapter put fields in the prebid extension is not as easy as it sounded. For now forget about my third point, we can just keep it how it is now parsing duration from the appnexus formatted bidder extension and just always look up the IAB category from the bid response to get the value for the hb_pb_cat_dur key. We can eventually figure out a uniform way for the adapters to pass through the category and duration video information for this key in the future.